lib/SIL/GraphFunctionDeviceInfo.cpp

```diff
@@ -14,21 +14,69 @@
 //
 //===----------------------------------------------------------------------===//
 
-
 #include "swift/SIL/GraphFunctionDeviceInfo.h"
 #include "swift/AST/DiagnosticsSIL.h"
 #include "swift/SIL/GraphOperationBuilder.h"
 #include "swift/SIL/GraphOperationInfo.h"
-#include "swift/SIL/SILInstruction.h"
 #include "swift/SIL/SILFunction.h"
+#include "swift/SIL/SILInstruction.h"
 #include "swift/SILOptimizer/Utils/Local.h"
-#include "llvm/ADT/StringRef.h"
+#include "tensorflow/c/c_api_experimental.h"
 #include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
 #include "llvm/Support/CommandLine.h"
+#include <iostream>
+
+#include <unordered_map>
 
 using namespace swift;
 using namespace tf;
 
+bool CanRunOnDevice(llvm::StringRef opType,
+                    llvm::ArrayRef<GraphOperationAttribute> attributes,
+                    const char *device_type) {
+  auto *builder = TF_NewAttrBuilder(std::string(opType).c_str());
+
+  unsigned dtypeAttr = 0;
+  for (const auto &attr : attributes) {
```

Comment:
what if `attributes` have 2 or more attrs of type `GraphOperationInfo::ArgumentLowering::TFDataTypeAttribute` -- do we want to assert that won't happen for now?

in that case it's not clear which type attr(s) we should use to query the op kernel database.

Reply:
Maybe this is just a cop-out, but these are the rules used to construct the ops themselves. Thus, we'll use the exact same attributes that will be used when actually running the op. I think if we want this to be stricter, we should make the actual op-checking be stricter as well.

Reply:
I think it'll already work with 2 or more TFDataTypeAttribute attrs. The calls to `TF_AttrBuilderSetType` and `TF_AttrBuilderSetTypeList` pass the attr name, so the attrbuilder should be able to keep track of which value corresponds to which attr.

Reply:
SG. From additional discussion with Parker today, it might be useful to have some compiler unit tests, that show for a given op:
- in one case, it's placed on GPU
- but when we change the data type a bit, it's placed on CPU.

Parker can decide whether to add that in this patch or later.